service/history/task/task.go

```diff
@@ -447,6 +454,21 @@ func (t *taskImpl) shouldResubmitOnNack() bool {
 		(t.queueType == QueueTypeActiveTransfer || t.queueType == QueueTypeActiveTimer)
 }
 
+func (t *taskImpl) shouldDropTaskOnError() bool {
+	if t.shard.GetConfig().EnableDropStuckTaskByDomainID(t.Info.GetDomainID()) && // use domainID here to avoid accessing domainCache
+		t.GetAttempt() > t.criticalRetryCount()/2 {
+		if transferTask, ok := t.Info.(*persistence.TransferTaskInfo); ok &&
+			transferTask.TaskType == persistence.TransferTaskTypeCloseExecution {
+			return true
+		}
+		if timerTask, ok := t.Info.(*persistence.TimerTaskInfo); ok &&
+			timerTask.TaskType == persistence.TaskTypeWorkflowTimeout {
```

Comment:
Can you elaborate more on why you think this task needs/can be dropped?

Note that workflow close/timeout isn't a low-priority event. If we fail to close a workflow, users won't be able to reuse the workflow ID and have other issues. Therefore, we should treat workflow close/timeout events as high priority as start workflow requests.

Reply:
Most of the stuck tasks are close/timeout tasks. 

> If we fail to close a workflow, users won't be able to reuse the workflow ID and have other issues.

We check the error to make sure the workflow is a resurrected workflow.